File: _book/src/main/python/com/frankcooper/bank/_19.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
    dummy = ListNode(-1)  # 定义一个哑结点
    dummy.next = head  ##将哑结点放在链表的头部
    slow, fast = dummy, dummy  ## 初始化时，快慢指针都指向哑结点
    for i in range(n):  # 调整快指针的位置，使其与慢指针的距离相差n
        fast = fast.next
    while fast.next:  # 游走快慢指针，当快指针走到链表的末尾时，慢指针的next节点指向的是待删除的节点
        slow = slow.next
        fast = fast.next
    slow.next = slow.next.next  # 开始删除
    return dummy.next  # 返回头节点

File: src/main/java/com/frankcooper/platform/leetcode/bank/_1_100/_19.java
package com.frankcooper.platform.leetcode.bank._1_100;

import com.frankcooper.io.ListNodeIOUtils;
import com.frankcooper.struct.ListNode;

/*import java.util.*;
import org.junit.Assert;*/
public class _19 {

    static class _1st {
        public static void main(String[] args) {
            _1st handler = new _1st();
            ListNode head = ListNodeIOUtils.transform("[1,2,3,4,5]");
            int n = 2;
            handler.removeNthFromEnd(head, n);

        }

        public ListNode removeNthFromEnd(ListNode head, int n) {
            ListNode dummy = new ListNode(-1);
            dummy.next = head;
            ListNode fast = dummy;
            for (int i = 0; i <= n; i++) fast = fast.next;
            ListNode slow = dummy;
            while (fast != null) {
                slow = slow.next;
                fast = fast.next;
            }
            slow.next = slow.next.next;
            return dummy.next;
        }

    }

    static class _2nd {
        public static void main(String[] args) {
            _2nd handler = new _2nd();
        }
    }


    static class _3rd {
        public static void main(String[] args) {
            _3rd handler = new _3rd();
        }
    }

    static class _4th {
        public static void main(String[] args) {
            _4th handler = new _4th();
        }
    }
}
